### PR TITLE
network: log traffic and add a simple traffic analysis script

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

**Comment by moneromooo-monero** [comment]
Tests build fixed.

File: src/cryptonote_protocol/levin_notify.cpp [hunk 1]
```diff
@@ -275,19 +283,23 @@ namespace levin
...
              making things linkable across connections. */
           if (this->source_ != context.m_connection_id && (this->zone_->is_public || !context.m_is_income))
-            connections.emplace_back(context.m_connection_id);
+            connections.emplace_back(context);
           return true;
         });
 
-        for (const boost::uuids::uuid& connection : connections)
-          zone_->p2p->send(message_.clone(), connection);
+        std::string category = "command-" + std::to_string(get_command_from_message(message_));
```
Comment (vtnerd) [inline src/cryptonote_protocol/levin_notify.cpp]: The result of `get_command_from_message` can be given directly to `on_levin_traffic` now.

File: contrib/epee/include/storages/levin_abstract_invoke2.h [hunk 2]
```diff
@@ -34,10 +34,26 @@
...
+  template<typename context_t>
+  void on_levin_traffic(const context_t &context, bool initiator, bool sent, bool error, size_t bytes, const char *category)
+  {
+    MCINFO("net.p2p.traffic", context << bytes << " bytes " << (sent ? "sent" : "received") << (error ? "/corrupt" : "")
+        << " for category " << category << " initiated by " << (initiator ? "us" : "peer"));
+  }
+  template<typename context_t>
+  void on_levin_traffic(const context_t &context, bool initiator, bool sent, bool error, size_t bytes, int command)
+  {
+    return on_levin_traffic(context, initiator, sent, error, bytes, ("command-" + std::to_string(command)).c_str());
```
Comment (vtnerd) [inline contrib/epee/include/storages/levin_abstract_invoke2.h]: I think its preferable to not allocate a `std::string` here for this - keep cycles to a minimum with optional debugging stuff.

The levin protocol uses a `std::uint32_t` value command. So a negative command value is illegal. And noise uses "0", which is currently a unique value (I chose it for that reason). So negative values for invalid commands and zero for noise could be used to drop the `const char*` overload entirely.

File: contrib/epee/include/storages/levin_abstract_invoke2.h [hunk 3]
```diff
@@ -34,10 +34,26 @@
 #undef MONERO_DEFAULT_LOG_CATEGORY
 #define MONERO_DEFAULT_LOG_CATEGORY "net"
 
+namespace
+{
+  template<typename context_t>
+  void on_levin_traffic(const context_t &context, bool initiator, bool sent, bool error, size_t bytes, const char *category)
+  {
+    MCINFO("net.p2p.traffic", context << bytes << " bytes " << (sent ? "sent" : "received") << (error ? "/corrupt" : "")
```
Comment (vtnerd) [inline contrib/epee/include/storages/levin_abstract_invoke2.h]: I meant to comment on this too - this level seems too low. We're writing to the log constantly just for bandwidth tracking stuff. If its that important, having the information tracked internally and then dumped periodically (either to log or elsewhere) should be a lot faster.

File: src/cryptonote_protocol/levin_notify.cpp [hunk 4]
```diff
@@ -275,19 +283,23 @@ namespace levin
...
+        for (const detail::p2p_context &context : connections)
+        {
+          on_levin_traffic(context, true, true, false, message_.size(), category.c_str());
```
Comment (vtnerd) [inline src/cryptonote_protocol/levin_notify.cpp]: Also, this should be moved into the `foreach_connection` above. Copying the entire context is going to be expensive, especially after the Dandelion++ fluff changes. There isn't a way to implement that without having a per-connection tx buffer.

File: contrib/epee/include/storages/levin_abstract_invoke2.h (hunk 3 above)
Comment (moneromooo-monero) [reply contrib/epee/include/storages/levin_abstract_invoke2.h]: It is not on by default.

**Pull request merged**